Create a DNAMutation class to mutate DNA sequences

Paucey opened the pull request:
Hi, first time contributor to this project and wanted to help. I took a look at issue #457 and came up with this.

Introduces a DNAMutation class which handles generating mutated DNA sequences. Specifically:

Mutate a DNA sequence by randomly changing a specified number of bases.
Generate 10 mutated sequences and print them to the console.

> generateMutatedSequences(): Given a DNA sequence generates 10 mutated sequences, each with a specified number of mutations, and prints them to the console.

> mutate(): Applies mutations by randomly selecting positions in the given DNA sequence and changing the base at that position.

> getDifferentBase(): Ensures the mutation results in a base different from the given original base.

Not sure if this was specifically what you were looking for (and it isn't integrated any further than just being the abstract methods) but again, wanted to contribute. Let me know if you would like anything tweaked.

Fixes #457.

Paucey commented:
Currently mutate() simply replaces one base in a sequence with another base, but it could be refined to offer more specific kinds of mutations.

VerisimilitudeX commented:
Hi @Paucey, thank you so much for your contribution! Taking a preliminary glance-through of your changes, everything looks good. I’ll look over it more in-depth over the weekend and will come back to you with feedback!

Paucey commented:
I'm admittedly not very knowledgeable when it comes to the science or technicalities of DNA mutation, so this may be the extent that my knowledge can take me re: a DNA Mutation class. If there is anything you'd like me to add or change, though, I can do my best.

VerisimilitudeX reviewed (changes requested):
@Paucey, your code looks great and the functionality is correct! Great work. However, this needs to be integrated in the CmdArgs file as a new command-line parameter. It should be fairly simple to implement as you would just have to clone the existing template.

Let me know if you need any help!

VerisimilitudeX commented:
I’ll approve it as soon as that’s done!

Paucey commented:
What do you think of this latest commit, re: CmdArgs?

I wasn't sure where to put the call to DNAMutation, so I snuck it in after the other DNAnalyzer print outs, and before the check/print out for if the DNA is random.

The only issue I can see with the current implementation is that there there's no upper limit on how many mutations to the DNA sequence one can request. I was thinking of two possible solutions for this -

- I could add a validator to run() in CmdArgs, checking the requested amount of mutations against the length of the DNA string, then either return an exception to the user and qutting, or print a message that the number of mutations requested exceeds the length of the string and set mutationCount to 0 (letting the rest of the program run normally, but skipping the mutationCount process later).
- Or, I could alter the mutate method in the DNAMutation class so that the number mutations is min()'d with the length of the given DNA string, defaulting to the length of the string if the requested number of mutations is too high.

Paucey commented:
Also, it occurred to me just now that in my current commit I am passing a DNAAnalysis object to mutate instead of a String...

VerisimilitudeX commented:
The update looks great! I think option 2 sounds the best! Probably just print a little warning acknowledging that. Thanks for being so proactive with this PR btw.

Also, just to confirm, if you run the program from Main.java or through Gradle using the Cmd arg mutation, does it run correctly and mutate the DNA sequence? Does it output it directly to the user or does it simply store it in the object?

Paucey commented:
Okay, made some more changes:

- Added the missing import for DNAMutation in CmdArgs
- Added 'N' to the bases array in DNAMutation to account for unknown bases in DNA
- Renamed "generateMutatedSequences()" to "generateAndPrintMutatedSequences()" to provide clarity of its function
- Removed a comma that was causing a syntax error
- Added a validator in DNAMutation.mutate() that checks if the requested number of mutations exceeds the length of the given DNA string, and if so, prints a warning and limits the mutations to the length of the string

As it stands, it prints the mutations to terminal rather than storing them in an object. Since the argument requires a value to run, it won't do nothing, as it stands, with the current implementation of Main unless the mutationCount value is manually altered to be greater than 0 in CmdArgs.java. Having tested what that looks like when that number *is* altered, it currently outputs... a lot.

Paucey commented:
Let me know how you'd like me to proceed - I could move the DNAMutation logic to DNATools, refactor the logic to be more efficient with the DNA strings, and could store the mutated DNA sequences in an object/objects?

If I were to store the mutated strings as objects, what should be done with them at that point? Written to a file/files?

VerisimilitudeX commented:
Would you like to create separate PRs for these? It'd help in terms of repository best-practices and also give you more visibility for your excellent quality contributions.

> DNAMutation logic to DNATools, refactor the logic to be more efficient with the DNA strings, and could store the mutated DNA sequences in an object/objects

VerisimilitudeX commented:
Also, if you could resolve the minor conflicts that'd be great!

VerisimilitudeX reviewed (changes requested):
Commented with feedback!

Paucey commented:
Okay, so I updated my branch to be in line with the main branch, and changed the generate-and-print-mutation method to now be a generate-and-write-mutation method, writing the 10 generated mutations to a .FA file (which is dynamically named with a time-stamp). Each mutation is identified in the file by a definition line (">mutation_" followed by the number that the mutation is), and the DNA strings are wrapped to 80 characters per line which, if I understand correctly, is standard for FASTA files.

I also moved the mutate call in CmdArgs to be before the other calls so that it's the original DNA that is mutated, not the reversed or reverse complemented, etc.

I have some other pull requests waiting in the wings - to move the mutation functionality to DNATools, to store the mutated DNA sequences in an object, and some minor improvements for some of the mutation functions. The last big one, which I haven't started yet, will be to refactor everything to be more efficient when working with the large DNA strings.

VerisimilitudeX reviewed (approved):
Awesome, @Paucey! Thank you so much for making so many modifications. This looks ready to merge - congratulations!

VerisimilitudeX commented:
Looking forward to seeing your other PRs come through soon!
